packages/webpack/src/BuildMetadataPlugin.ts

```diff
@@ -0,0 +1,118 @@
+import { execSync } from 'node:child_process';
+import * as path from 'node:path';
+import { PackageJson } from 'type-fest';
+import { type WebpackPluginInstance, BannerPlugin, Compilation, Compiler, sources } from 'webpack';
+
+const getBaseBuildMetadata = () => {
+  const now = new Date();
+
+  return {
+    buildDateTime: now.toISOString(),
+    gitCommit: execSync('git rev-parse HEAD').toString().trim(),
+    gitBranch: execSync('git rev-parse --abbrev-ref HEAD').toString().trim(),
+    gitTags: execSync('git tag --points-at HEAD').toString().trim().split('\n').filter(Boolean),
+  };
+};
+
+export type BuildMetadataPluginOptions = {
+  /**
+   * The contents of `package.json` to provide package name and version.
+   */
+  packageJson: PackageJson;
+
+  /**
+   * Output path for the metadata file.
+   *
+   * Defaults value: the configured webpack output path
+   */
+  outputPath?: string;
+
+  /**
+   * Name of the JSON metadata file.
+   *
+   * Default value: `build-metadata.json`
+   */
+  jsonMetadataFilename?: string;
+
+  /**
+   * The tag to include in the chunk banner.  This can be used to configure terser to
+   * retain the banner comment in the optimized chunks.
+   *
+   * Default value: `@buildInfo`
+   */
+  bannerBuildInfoTag?: string;
+};
+
+/**
+ * Collect and output build metadata from a provided `package.json` and from calls to
+ * git.  The metadata is saved in two places:
+ *   - A banner comment starting with `bannerBuildInfoTag` in each chunk
+ *   - A json metadata file
+ */
+export class BuildMetadataPlugin implements WebpackPluginInstance {
+  private readonly packageJson: PackageJson;
+  private readonly outputPath: string;
+  private readonly jsonMetadataFilename: string;
+  private readonly bannerBuildInfoTag: string;
+
+  constructor(options: BuildMetadataPluginOptions = { packageJson: {} }) {
+    this.packageJson = options.packageJson;
+    this.outputPath = options.outputPath ?? '.';
+    this.jsonMetadataFilename = options.jsonMetadataFilename ?? 'build-metadata.json';
+    this.bannerBuildInfoTag = options.bannerBuildInfoTag ?? '@buildInfo';
+  }
+
+  apply(compiler: Compiler) {
+    //
+    // Lookup the build metadata (package.json of the thing running webpack & git info):
+    //
+    const buildInfo = {
+      packageName: this.packageJson.name,
+      packageVersion: this.packageJson.version,
+      ...getBaseBuildMetadata(),
+    };
+
+    //
+    // Build up the chunk banner (include the @buildInfo tag so terser can be configured to keep it):
+    //
+    const bannerInfo = [
```

Comment:
same as the comment about the build info, can we use one central method to create the banner string for both rollup and webpack, and consume it here ?

p.s I prefer the simplicity of just using the same Metadata as the we use in the build Metadata file

Reply:
The banners need to be a little bit different so terser can be configured to retain the metadata comments, but I'll make them very similar.